### PR TITLE
Rename `TypedNode.whole` into `program`

### PR DESCRIPTION
**Pull request by lucteo**

### CODE REVIEW

File: Sources/Compiler/TypedNode.swift [hunk 1]
```diff
@@ -13,10 +13,10 @@ extension AST {
 @dynamicMemberLookup
 public struct TypedNode<ID: NodeIDProtocol> : Hashable {
 
-  /// The whole program of which this node is a notional part.
-  fileprivate let whole: TypedProgram
+  /// The program program of which this node is a notional part.
+  fileprivate let program: TypedProgram
```
[the remark this answers is not in the document]
Comment (lucteo) [reply Sources/Compiler/TypedNode.swift:17]: @dabrahams : suggested to me to make this `fileprivate` after I abused it several times to get the typed program from a typed node

[the remark this answers is not in the document]
Comment (kyouko-taiga) [reply Sources/Compiler/TypedNode.swift:17]: Interesting! Well then I guess it confirms my suspicion then. Let's keep it private.

[the remark this answers is not in the document]
Comment (lucteo) [reply Sources/Compiler/TypedNode.swift:17]: It needs to be `fileprivate`, not `private`. Making it `private` will result in all the places we construct TypedNodes in this file.

[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/Compiler/TypedNode.swift:17]: It's not reference semantics if there's no mutation.

[the remark this answers is not in the document]
Comment (kyouko-taiga) [reply Sources/Compiler/TypedNode.swift:17]: @lucteo Yes, `fileprivate`. I only meant "not public".

@dabrahams Yes, I meant "in a world with reference semantics, this pattern is a reference hub".

[the remark this answers is not in the document]
Comment (dabrahams) [reply Sources/Compiler/TypedNode.swift:17]: @kyouko-taiga you asked what I thought; that's the only thought I have.

File: Sources/Compiler/TypedNode.swift [hunk 2]
```diff
@@ -38,15 +38,15 @@ extension TypedProgram {
   /// Bundles `id` together with `self`.
   public subscript<TargetID: NodeIDProtocol>(_ id: TargetID) -> TypedNode<TargetID>
   {
-    TypedNode<TargetID>(whole: self, id: id)
+    TypedNode<TargetID>(program: self, id: id)
```
Comment (dabrahams) [inline Sources/Compiler/TypedNode.swift:41]:
```suggestion
    TypedNode(program: self, id: id)
```

Comment (dabrahams) [reply Sources/Compiler/TypedNode.swift:41]: seeing this makes me think we should have a different initializer signature, allowing:

```swift
    TypedNode(id, in: program)
```

**Review by dabrahams — APPROVED** [review]
Nice; see comments inline.

File: Sources/Compiler/TypedNode.swift (hunk 2 above)
Comment (lucteo) [reply Sources/Compiler/TypedNode.swift]: This makes more sense

**Pull request merged**